File: elixir/protein-translation/protein_translation.exs
defmodule ProteinTranslation do
  @rna_error "invalid RNA"
  @codon_error "invalid codon"

  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {atom, list(String.t())}
  def of_rna(rna) do
    of_rna(rna, [])
  end

  defp of_rna("", acc) do
    {:ok, Enum.reverse(acc)}
  end

  defp of_rna(<<codon1 :: utf8, codon2 :: utf8, codon3 :: utf8, rna :: binary>>, acc) do
    codon = <<codon1, codon2, codon3>>
    case of_codon(codon) do
      {:error, _message} ->
        {:error, @rna_error}
      {:ok, "STOP"} ->
        of_rna("", acc)
      {:ok, protein} ->
        of_rna(rna, [protein | acc])
    end
  end

  defp of_rna(_rna, _acc) do
    {:error, @rna_error}
  end

  @doc """
  Given a codon, return the corresponding protein

  UGU -> Cysteine
  UGC -> Cysteine
  UUA -> Leucine
  UUG -> Leucine
  AUG -> Methionine
  UUU -> Phenylalanine
  UUC -> Phenylalanine
  UCU -> Serine
  UCC -> Serine
  UCA -> Serine
  UCG -> Serine
  UGG -> Tryptophan
  UAU -> Tyrosine
  UAC -> Tyrosine
  UAA -> STOP
  UAG -> STOP
  UGA -> STOP
  """
  @spec of_codon(String.t()) :: {atom, String.t()}
  def of_codon("UGU"), do: {:ok, "Cysteine"}
  def of_codon("UGC"), do: {:ok, "Cysteine"}
  def of_codon("UUA"), do: {:ok, "Leucine"}
  def of_codon("UUG"), do: {:ok, "Leucine"}
  def of_codon("AUG"), do: {:ok, "Methionine"}
  def of_codon("UUU"), do: {:ok, "Phenylalanine"}
  def of_codon("UUC"), do: {:ok, "Phenylalanine"}
  def of_codon("UCU"), do: {:ok, "Serine"}
  def of_codon("UCC"), do: {:ok, "Serine"}
  def of_codon("UCA"), do: {:ok, "Serine"}
  def of_codon("UCG"), do: {:ok, "Serine"}
  def of_codon("UGG"), do: {:ok, "Tryptophan"}
  def of_codon("UAU"), do: {:ok, "Tyrosine"}
  def of_codon("UAC"), do: {:ok, "Tyrosine"}
  def of_codon("UAA"), do: {:ok, "STOP"}
  def of_codon("UAG"), do: {:ok, "STOP"}
  def of_codon("UGA"), do: {:ok, "STOP"}
  def of_codon(_), do: {:error, @codon_error}
end

File: erlang/protein-translation/src/protein_translation.erl
-module(protein_translation).

-export([proteins/1, test_version/0]).

-type nucleotides() :: [$A | $C | $G | $T].
-type proteins() :: [cysteine | leucine | methionine | phenylalanine | serine | tryptophan | tyrosine].

%% API

-spec proteins(nucleotides()) -> proteins().
proteins(Strand) ->
    proteins(Strand, []).

-spec test_version() -> integer().
test_version() ->
    1.

%% Internal

proteins([], Acc) ->
    lists:reverse(Acc);
proteins([N1, N2, N3 | Strand], Acc) ->
    case protein([N1, N2, N3]) of
        stop ->
            proteins([], Acc);
        Protein ->
            proteins(Strand, [Protein | Acc])
    end.

protein("UGU") -> cysteine;
protein("UGC") -> cysteine;
protein("UUA") -> leucine;
protein("UUG") -> leucine;
protein("AUG") -> methionine;
protein("UUU") -> phenylalanine;
protein("UUC") -> phenylalanine;
protein("UCU") -> serine;
protein("UCC") -> serine;
protein("UCA") -> serine;
protein("UCG") -> serine;
protein("UGG") -> tryptophan;
protein("UAU") -> tyrosine;
protein("UAC") -> tyrosine;
protein("UAA") -> stop;
protein("UAG") -> stop;
protein("UGA") -> stop.
